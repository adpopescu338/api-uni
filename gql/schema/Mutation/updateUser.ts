import { gql } from 'graphql-tag';

export const updateUser = gql`
  type Mutation {
    """
    If you're not sysadmin, you must be authenticated as the user you're trying to update
    """
    updateUser(id: ID!, input: UserInput!): User!
  }
`;
