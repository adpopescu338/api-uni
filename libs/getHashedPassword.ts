import bcrypt from 'bcryptjs';

export async function getHashedPassword(password: string) {
  const salt = await bcrypt.genSalt(10);
  return await bcrypt.hash(password, salt);
}
