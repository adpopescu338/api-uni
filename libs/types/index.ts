export type { Context } from './Context';
export * from './generated';
export { GQLResolver } from './GqlResolver';
export { RedisSession } from './RedisSession';
