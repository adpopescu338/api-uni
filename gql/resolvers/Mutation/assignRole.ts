import { Context } from 'libs/types';
import { MutationAssignRoleArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from '../../../libs/throwIfNotAdmin';

export const assignRole = async (
  _: unknown,
  { userId, roleId }: MutationAssignRoleArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const result = await prisma.user.update({
    where: {
      id: userId,
    },
    data: {
      roles: {
        connect: {
          id: roleId,
        },
      },
    },
    include: {
      roles: {
        include: {
          permissions: true,
        },
      },
    },
  });

  return result;
};
