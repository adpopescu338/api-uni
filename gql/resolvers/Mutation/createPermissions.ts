import { Context } from 'libs/types';
import { MutationCreatePermissionsArgs } from 'libs/types/generated';

export const createPermissions = async (
  _: unknown,
  { permissions }: MutationCreatePermissionsArgs,
  { prisma }: Context,
) => {
  const createdPermissions = await prisma.permission.createMany({
    data: permissions,
  });

  return createdPermissions;
};
