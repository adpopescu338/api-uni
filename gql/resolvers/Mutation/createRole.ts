import { Context } from 'libs/types';
import { MutationCreateRoleArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from 'libs/throwIfNotAdmin';

export const createRole = async (
  _: unknown,
  { name, description, permissionIds }: MutationCreateRoleArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const role = await prisma.userDefinedRole.create({
    data: {
      name,
      description,
      ...(permissionIds.length && {
        permissions: {
          connect: permissionIds.map((id) => ({ id })),
        },
      }),
    },
  });

  return role;
};
