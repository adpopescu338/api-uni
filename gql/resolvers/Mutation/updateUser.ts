import { Context } from 'libs/types';
import { MutationUpdateUserArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from '../../../libs/throwIfNotAdmin';
import { getHashedPassword } from '../../../libs/getHashedPassword';

export const updateUser = async (
  _: unknown,
  { input, id }: MutationUpdateUserArgs,
  { prisma, user }: Context,
) => {
  // validate that the user is either sysadmin or the user being updated
  if (id !== user?.id) {
    throwIfNotAdmin(user);
  }

  const password = input.password ? await getHashedPassword(input.password) : undefined;

  const updatedUser = await prisma.user.update({
    where: {
      id,
    },
    data: {
      ...input,
      password,
      isSysAdmin: input.isSysAdmin || false,
    },
  });

  return updatedUser;
};
