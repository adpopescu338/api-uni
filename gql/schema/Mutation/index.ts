import { logout } from './logout';
import { login } from './login';

import { createRole } from './createRole';
import { updateRole } from './updateRole';
import { deleteRole } from './deleteRole';
import { assignRole } from './assignRole';

import { createPermissions } from './createPermissions';
import { updatePermissions } from './updatePermissions';
import { deletePermission } from './deletePermission';

import { createUser } from './createUser';
import { updateUser } from './updateUser';

export const Mutation = [
  login,
  logout,

  createRole,
  updateRole,
  deleteRole,
  assignRole,

  createPermissions,
  updatePermissions,
  deletePermission,

  createUser,
  updateUser,
];
