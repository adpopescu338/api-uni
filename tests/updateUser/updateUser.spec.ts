import { describe, it, expect, jest } from '@jest/globals';
import { updateUser } from '../../gql/resolvers/Mutation/updateUser';
import { Context } from 'libs/types';
import * as getHashedPasswordModule from 'libs/getHashedPassword';

jest.spyOn(getHashedPasswordModule, 'getHashedPassword').mockResolvedValue('hashed--password');

const expectedResult = {
  id: 'id',
  name: 'name',
  email: 'email',
  password: 'hashed--password',
  roles: [],
};

const ctx = {
  prisma: {
    user: {
      update: jest.fn().mockResolvedValue(expectedResult as never),
    },
  },
  user: {
    isSysAdmin: true,
    id: 'id',
  },
  cacheManager: {
    del: jest.fn(),
    cacheKeys: {
      usersCount: 'usersCount',
    },
  },
} as unknown as Context;

const payload = {
  input: {
    name: 'name',
    email: 'email',
    password: 'password',
  },
  id: ctx.user!.id,
};

describe('updateUser', () => {
  it('Should be defined', () => {
    expect(updateUser).toBeDefined();
  });

  it('Should throw an error if user is not system admin and the userId is not the one provided', async () => {
    await expect(
      updateUser(undefined, payload, {
        ...ctx,
        user: { isSysAdmin: false, id: 'abc' },
      } as Context),
    ).rejects.toThrowError();
  });

  it('Should save the correct data and return the user', async () => {
    const result = await updateUser(undefined, payload, ctx);
    const { lastCall } = (ctx.prisma.user.update as jest.Mock).mock;

    expect(lastCall).toMatchSnapshot();

    expect(result).toEqual(expectedResult);
  });
});
