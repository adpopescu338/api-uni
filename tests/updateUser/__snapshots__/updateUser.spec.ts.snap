// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateUser Should save the correct data and return the user 1`] = `
[
  {
    "data": {
      "email": "email",
      "isSysAdmin": false,
      "name": "name",
      "password": "hashed--password",
    },
    "where": {
      "id": "id",
    },
  },
]
`;
